pkg/rhos/securitygroups.go

```diff
@@ -0,0 +1,244 @@
+		}
+
+		return true, nil
+	})
+
+	return errors.WithMessage(err, "failed to removed security group from servers")
+}
+
+func (c *CloudInfo) createGWSecurityGroup(ports []api.PortSpec, groupName string, computeClient *gophercloud.ServiceClient,
+	networkClient *gophercloud.ServiceClient) error {
+	isFound, err := checkIfSecurityGroupPresent(groupName, computeClient)
+	if err != nil {
+		return err
+	}
+
+	if isFound {
+		return nil
+	}
+
+	opts := secgroups.CreateOpts{
+		Name:        groupName,
+		Description: "Submariner Gateway",
+	}
+
+	group, err := secgroups.Create(computeClient, opts).Extract()
+	if err != nil {
+		return errors.WithMessage(err, "failed to create g/w security group")
+	}
+
+	for _, port := range ports {
+		err = c.createSGRule(group.ID, "", allNetworkCIDR, port.Port, port.Protocol, networkClient)
+		if err != nil {
+			return errors.WithMessagef(err, "creating security group rule failed")
+		}
+	}
+
+	return nil
+}
+
+func checkIfSecurityGroupPresent(groupName string, computeClient *gophercloud.ServiceClient) (bool, error) {
+	pager := secgroups.List(computeClient)
+	var isFound bool
+
+	err := pager.EachPage(func(page pagination.Page) (bool, error) {
+		serverList, err := secgroups.ExtractSecurityGroups(page)
+
+		for _, s := range serverList {
+			if s.Name == groupName {
+				isFound = true
+			}
+		}
+		return true, errors.WithMessagef(err, "failed to extract the security group %q from results", groupName)
+	})
+
+	return isFound, errors.WithMessagef(err, "error getting the security group : %q", groupName)
+}
+
+func (c *CloudInfo) openGatewayPort(groupName, nodeName string, computeClient *gophercloud.ServiceClient) error {
+	opts := servers.ListOpts{Name: nodeName}
+	pager := servers.List(computeClient, opts)
+
+	err := pager.EachPage(func(page pagination.Page) (bool, error) {
+		serverList, err := servers.ExtractServers(page)
+		if err != nil {
+			return false, errors.WithMessagef(err, "getting the server List failed for node %q", nodeName)
+		}
+		for i := range serverList {
+			err = secgroups.AddServer(computeClient, serverList[i].ID, groupName).ExtractErr()
+			if err != nil {
+				return false, errors.WithMessagef(err, "adding security group %q to the server %q failed",
+					groupName, serverList[i].Name)
+			}
+		}
+
+		return true, nil
+	})
+
+	return errors.WithMessagef(err, "open gateway ports failed")
+}
+
+func (c *CloudInfo) removeGWFirewallRules(groupName, nodeName string, computeClient *gophercloud.ServiceClient) error {
+	opts := servers.ListOpts{Name: nodeName}
+	pager := servers.List(computeClient, opts)
+
+	err := pager.EachPage(func(page pagination.Page) (bool, error) {
+		serverList, err := servers.ExtractServers(page)
+		if err != nil {
+			return false, errors.WithMessagef(err, "getting the server list failed")
+		}
+		for i := range serverList {
+			err = secgroups.RemoveServer(computeClient, serverList[i].ID, groupName).ExtractErr()
```

Comment:
What if the SG rule is missing? Should we check the return code to see if its NotFound and avoid returning error in such cases?

Reply:
Checking for the exact error code now.